presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -921,9 +958,11 @@ public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup,
             super(Optional.empty(), name, jmxExportListener, executor);
         }
 
-        public synchronized void processQueuedQueries()
+        public synchronized void updateGroupsAndProcessQueuedQueries()
         {
-            internalRefreshStats();
+            updateResourceUsageAndGetDelta();
```

Comment:
Maybe I don't understand this, but if updateResourceUsageAndGetDelta is a recursive function and internalRefreshStats is not, then I believe this will add greatly to how CPU-intensive this call is.

Reply:
`internalRefreshStats` and `updateResourceUsageAndGetDelta` are both recursive. One difference is in how they update the usage. The usage delta in `updateResourceUsageAndGetDelta` is returned from the stack through a child call.  `internalRefreshStats` stores the updated usage on the heap, the parent picks it up and does the delta calculation itself.